gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/internal/ComposeCompilerArtifactProvider.kt

```diff
@@ -5,33 +5,69 @@
 
 package org.jetbrains.compose.internal
 
-import org.jetbrains.compose.ComposeBuildConfig
+import org.jetbrains.compose.ComposeCompilerCompatability
+import org.jetbrains.kotlin.gradle.plugin.KotlinPlatformType
+import org.jetbrains.kotlin.gradle.plugin.KotlinTarget
 import org.jetbrains.kotlin.gradle.plugin.SubpluginArtifact
 
-internal class ComposeCompilerArtifactProvider(private val customPluginString: () -> String?) {
+private const val KOTLIN_COMPATABILITY_LINK =
+    "https://github.com/JetBrains/compose-jb/blob/master/VERSIONING.md#kotlin-compatibility"
+
+internal class ComposeCompilerArtifactProvider(
+    private val kotlinVersion: String,
+    customPluginString: () -> String?
+) {
+    fun checkTargetSupported(target: KotlinTarget) {
+        if (target.platformType == KotlinPlatformType.js) {
+            require(isJsSupported) {
+                "This version of Compose Multiplatform doesn't support Kotlin " +
+                    "$kotlinVersion for ${target.platformType} target. " +
+                    "Please see $KOTLIN_COMPATABILITY_LINK " +
+                    "to know the latest supported version of Kotlin."
+            }
+        }
+    }
+
+    private var isJsSupported = true
+
     val compilerArtifact: SubpluginArtifact
-        get() {
-            val customPlugin = customPluginString()
-            val customCoordinates = customPlugin?.split(":")
-            return when (customCoordinates?.size) {
-                null -> DefaultCompiler.pluginArtifact
-                1 -> {
-                    val customVersion = customCoordinates[0]
-                    check(customVersion.isNotBlank()) { "'compose.kotlinCompilerPlugin' cannot be blank!" }
-                    DefaultCompiler.pluginArtifact.copy(version = customVersion)
+
+    init {
+        val customPlugin = customPluginString()
+        val customCoordinates = customPlugin?.split(":")
+        when (customCoordinates?.size) {
+            null -> {
+                val version = requireNotNull(
+                    ComposeCompilerCompatability.compilerVersionFor(kotlinVersion)
+                ) {
+                    "This version of Compose Multiplatform doesn't support Kotlin " +
```

Comment:
The message is a bit confusing. As I understand it, here we simply don't know, what version of Compose Compiler to choose. Instead the message says "this version of Compose doesn't support Kotlin $kotlinVersion", which may or may not be true. For example, if there is a version of a Compose Compiler, that works with current plugin and the desired version of Kotlin, then in my understanding the Gradle plugin & libraries are compatible, we simply don't know the right version of the compiler.

Reply:
Also, there is an option to completely prohibit using Kotlin 1.7.20, even if the user sets compatible Compose Compiler. But I don't think it is a good idea.